libs/failsafe-controller/src/main/java/net/jodah/failsafe/internal/executions/SyncControlledExecution.java

```diff
@@ -0,0 +1,250 @@
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package net.jodah.failsafe.internal.executions;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import javax.annotation.Nullable;
+import net.jodah.failsafe.ControlledExecutionException;
+import net.jodah.failsafe.FailsafeController;
+import net.jodah.failsafe.FailsafeException;
+import net.jodah.failsafe.SyncFailsafe;
+import net.jodah.failsafe.SyncFailsafeConfigDelegater;
+import net.jodah.failsafe.function.CheckedRunnable;
+import net.jodah.failsafe.function.ContextualCallable;
+import net.jodah.failsafe.function.ContextualRunnable;
+import net.jodah.failsafe.internal.monitor.ThreadMonitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class defines an {@link SyncFailsafe} which is used to control and track a specific failsafe
+ * execution.
+ *
+ * @param <R> the result type
+ */
+public class SyncControlledExecution<R> extends SyncFailsafeConfigDelegater<R>
+    implements ControlledExecution<R> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SyncControlledExecution.class);
+
+  private static final String EXECUTION = "execution";
+
+  private final FailsafeController<R> controller;
+
+  /**
+   * Reference to the master {@link SyncFailsafe} where all requests for more execution should be
+   * delegated.
+   */
+  private final SyncFailsafe<R> master;
+
+  /**
+   * Keeps track of all threads that are currently allocated to help failsafe perform a specific
+   * execution. These threads are either asynchronously executing the retry attempt or
+   * asynchronously notifying registered listeners.
+   */
+  private final ThreadMonitor monitor;
+
+  /** Unique identifier for this execution. */
+  private final int id;
+
+  /** Keeps track of when a given execution completed successfully or in error. */
+  private boolean completed = false;
+
+  /**
+   * Keeps track of the result for the execution. This will be <code>null</code> until the execution
+   * completes at which points it might remain <code>null</code> if the result returned is <code>
+   * null</code> or if the execution failed.
+   */
+  @Nullable private R result = null;
+
+  /**
+   * Keeps track of the error for the execution. This will be <code>null</code> until the execution
+   * completes at which point it will remain <code>null</code> if no error occurred.
+   */
+  @Nullable private Throwable error = null;
+
+  public SyncControlledExecution(FailsafeController<R> controller, SyncFailsafe<R> master, int id) {
+    super(master);
+    this.controller = controller;
+    this.master = master;
+    this.monitor = new ThreadMonitor(controller, id);
+    this.id = id;
+  }
+
+  @Override
+  public int getId() {
+    return id;
+  }
+
+  @Override
+  public boolean hasCompleted() {
+    synchronized (controller) {
```

Comment:
❓ It looks like a lot of the places calling methods are already synchronized, is there a reason we are doing it here as well.

Reply:
since it is part of the public interface for the `ControlledExecution`, I prefer not to rely on others. If this was a private method then I would be ok and actually also push for not synchronizing again.